lahja/asyncio/endpoint.py

```diff
@@ -494,68 +444,68 @@ def connect_to_endpoints_nowait(self, *endpoints: ConnectionConfig) -> None:
 
     async def connect_to_endpoint(self, config: ConnectionConfig) -> None:
         self._throw_if_already_connected(config)
-        if config.name in self._full_connections.keys():
-            self.logger.warning(
-                "Tried to connect to %s but we are already connected to that Endpoint",
-                config.name,
-            )
-            return
 
         conn = await Connection.connect_to(config.path)
-        remote = AsyncioRemoteEndpoint(config.name, conn, self._receiving_queue.put)
+        remote = AsyncioRemoteEndpoint(
+            local_name=self.name,
+            conn=conn,
+            subscriptions_changed=self._remote_subscriptions_changed,
+            new_msg_func=self._receiving_queue.put,
+        )
 
-        task = asyncio.ensure_future(self._handle_server(remote))
+        task = asyncio.ensure_future(self._run_remote_endpoint(remote))
 
         task.add_done_callback(self._endpoint_tasks.remove)
 
         self._endpoint_tasks.add(task)
 
         # block until we've received the other endpoint's subscriptions to
         # ensure that it is safe to broadcast
-        await remote.wait_until_subscription_received()
+        await self.wait_until_connected_to(config.name)
 
-        async with self._remote_connections_changed:
-            self._full_connections[config.name] = remote
-            self._remote_connections_changed.notify_all()
+    async def _run_remote_endpoint(self, remote: RemoteEndpointAPI) -> None:
+        await remote.start()
 
-    async def _handle_server(self, remote: RemoteEndpointAPI) -> None:
         try:
-            async with run_remote_endpoint(remote):
-                await remote.wait_stopped()
+            await remote.wait_ready()
+            await self._add_connection(remote)
+            await remote.wait_until_subscription_initialized()
+            await remote.wait_stopped()
         finally:
-            if remote.name is not None:
+            await remote.stop()
+            if remote in self._connections:
                 async with self._remote_connections_changed:
-                    self._full_connections.pop(remote.name)
+                    self._connections.remove(remote)
                     self._remote_connections_changed.notify_all()
+                async with self._remote_subscriptions_changed:
+                    self._remote_subscriptions_changed.notify_all()
```

Comment:
For symmetry with `self._add_connection` it might look better to pull this block out into a `self._remove_connection`.